### ISSUE TITLE
Hangfire/Background Job - Which layer do I implement it in?

### ISSUE THREAD

**Issue by tinonetic** [abpframework/abp#4174]
Hi,

In previous versions of ABP, we implemented Hangfire job class as a DomainService

We have 3 deployments working well.

I realise that we had it all wrong (or you have rethought the architecture?). 

After creating a new project using version 2.8.0, I am now seeing that what we used to do does not work anymore because of cyclic dependencies. In fact, Visual Studio won't allow the type of referencing we used to use.

I think our architecture worked before because we created a Dapper project to run store procedures (**Domain depends-on Dapper**). A similar approach breaks if depending on EF Core

Now that we are trying to use EF Core, we realise that **EF Core depends-on Domain**

I hope I made sense.

**TL;DR**
So my main questions are **in which layer do we place our Hangfire job?** and then **where do we place the cusotm Repository?** The docs do not say. [Only mentions creation of an Abp Module](https://docs.abp.io/en/abp/master/Background-Jobs-Hangfire)

https://docs.abp.io/en/abp/master/Best-Practices/Module-Architecture

https://docs.abp.io/en/abp/master/Background-Jobs-Hangfire

For the custom repository implementation, d[ocumentation **only mentions the repository interfaces.**](https://docs.abp.io/en/abp/master/Best-Practices/Module-Architecture#domain-layer) And it says they should be in the Domain Layer. Nothing about the implementation

As always, appreciate your awesome project!

**Comment by maliming** [comment]
> in which layer do we place our Hangfire job? 

The job can be placed in the application service layer or the domian layer according to the actual situation. The job class should depend on the interface rather than the specific implementation. This can avoid the problem of circular dependency.

> and then where do we place the cusotm Repository? 

The implementation of Repository is put in the EF Core layer or Mongodb layer.

**Comment by tinonetic** [comment]
Hi @maliming 

Thank-you for responding.

But I keep getting the error:

> None of the constructors found with 'Autofac.Core.Activators.Reflection.DefaultConstructorFinder' on type 'MyNamespace.Project.MyDomainService' can be invoked with the available services and parameters:
Cannot resolve parameter 'MyNamespace.Project.Repository.IMyEntityRepository myEntityRepository' of constructor 'Void .ctor(MyNamespace.Project.Repository.IMyEntityRepository)'

I have been stuck on this for 3days now.

This is what I did

> **MyNamespace.Project.EntityFrameworkCore**
>  - MyEntityRepository : EfCoreRepository<MyNamespaceDbContext, MyEntity, long>, IMyEntityRepository
> 
> **MyNamespace.Project.Domain**
> - IMyEntityRepository
> - MyDomainService : DomainService, ITransientDependency <--- this throws a DI exception


In the MyDomainServiceModule, I have

```c# 
        public override void ConfigureServices(ServiceConfigurationContext context)
        {
            Configure<AbpMultiTenancyOptions>(options =>
            {
                options.IsEnabled = MultiTenancyConsts.IsEnabled;
            });

            context.Services.AddTransient<IMyEntityRepository, MyEntityRepository>();
            context.Services.AddTransient<MyDomainService>();
        }

I have defined the DBSet<MyEntity>, EFCore migrations ran well.

Not sure what else I could be missing

Any help or guidance would be appreciated

**Comment by maliming** [comment]
hi @tinonetic  Can you share a project?

**Comment by tinonetic** [comment]
Hi @maliming . I will try and recreate a sample project exactly as we did

**Comment by tinonetic** [comment]
I have since resolved this issue, thanks
